pkg/server/registry/apigroups/acorn/volumes/policyswitcher.go

```diff
@@ -0,0 +1,49 @@
+package volumes
+
+import (
+	"context"
+
+	"github.com/acorn-io/mink/pkg/strategy"
+	"github.com/acorn-io/mink/pkg/strategy/remote"
+	"github.com/acorn-io/mink/pkg/types"
+	v1 "k8s.io/api/core/v1"
+)
+
+type PolicySwitcherStrategy struct {
+	s strategy.CompleteStrategy
+	t *Translator
+	r *remote.Remote
+}
+
+// NewPolicySwitcherStrategy returns a new policy switcher strategy that switch persistvolume policy from retain to delete when the volume is being deleted.
+// This makes sure the actual storage resource is deleted when the volume is deleted.
+func NewPolicySwitcherStrategy(s strategy.CompleteStrategy, t *Translator, r *remote.Remote) strategy.Deleter {
+	return PolicySwitcherStrategy{
+		t: t,
+		s: s,
+		r: r,
+	}
+}
+
+func (p PolicySwitcherStrategy) Delete(ctx context.Context, obj types.Object) (types.Object, error) {
+	pvName, err := p.t.FromVolumeToPVName(ctx, obj.GetNamespace(), obj.GetName())
+	if err != nil {
+		return nil, err
+	}
+	pvObj, err := p.r.Get(ctx, "", pvName)
+	if err != nil {
+		return nil, err
+	}
+	pv := pvObj.(*v1.PersistentVolume)
+	pv.Spec.PersistentVolumeReclaimPolicy = v1.PersistentVolumeReclaimDelete
+	_, err = p.r.Update(ctx, pv)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.s.Delete(ctx, obj)
```

Comment:
If this delete call fails we'd end up with a PV that doesn't have the retain policy but doesn't have a delete timestamp set.

I think we want to avoid getting into that state.

One suggestion to get around it: set the deletion timestamp before the update call to signal deletion.

Reply:
I think the behavior should be ok. If delete fails, we will just leave the PV's reclaimPolicy to delete and wait for user to issue another delete(which is always the case in manager because cluster-agent will try to remove it when upstream volume is gone). No real impact on the actual PV itself expect the reclaimpolicy but we will always modify that next time you call delete.